ADR for enhanced metadata generation (#40336)
quarkusio/quarkus: adr/0002-generate-extension-metadata-in-deployment-module.adoc

= Enhanced extension metadata generation

> * Status: proposed

Reviewer: Should be change on merge.

> * Date: 2024-04-26

Reviewer: Might need an update.

How do we know which extensions provide a dev service? At the moment, we don't, except by folk knowledge and https://quarkus.io/guides/dev-services[manually maintained lists.]

For further context on the need, see also

* https://github.com/quarkusio/quarkus/pull/40306
* https://github.com/quarkusio/quarkusio.github.io/issues/1869
* https://github.com/quarkusio/extensions/issues/983

When building an extension, we have an opportunity to introspect and surface a range of useful metadata, such as whether it provides a dev service.
But there's a problem. The introspection needs to happen in the deployment module, because that's where most of the interesting extension code is.
But the place we create the `quarkus-extension.yaml` is the runtime module. The deployment module depends on the runtime module, so it's a circular dependency.

Because the dependency isn't a code-compilation one, it doesn't break the build, but it does mean the generated metadata is only accurate on the second build.

image::images/0002-circular-dependency.svg[the circular dependency]
Source: https://excalidraw.com/#json=cTOOltCdWxadxC8QYc93R,AzxQWd7ALVSsk_FbyIuqAg

=== Related problems

> We have a similar circularity problem with config javadoc; the config is discovered in the deployment module, but the generated of the final-format data is done elsewhere. A solution here *might* help with that.

Reviewer 2: ```suggestion
We have a similar circularity problem with config javadoc; the config is discovered in the deployment module, but the generation of the final-format data is done elsewhere. A solution here *might* help with that.
```
Or am I wrong?

Although it's not a serious problem, it would also be nice to allow people to use `CapabilityBuildItems` to register capabilities.

> At the moment, if authors do that, the information https://quarkus.io/guides/capabilities#capabilitybuilditem[will not be available to the dev tools], so we need to have lots of warnings in our documentation telling them not to.

Reviewer 2: I added a comment about it https://github.com/quarkusio/quarkus/pull/40306#issuecomment-2235844459 too.
We may want to deprecate `CapabilityBuildItem` in the future and prefer the metadata-only approach.

Author: Yes, I can see the appeal of that. I'd say that we should either get rid of it, or make it work properly without warnings not to use it. :)

- *Doing nothing*. This is cheap, and non-breaking. But it means we miss an opportunity to take advantage of the fact that, at build time, we know a lot about what extensions can do and would like to make that information available to Quarkus users.
- *Move (most) metadata generation to the deployment module* We can eliminate the circular dependency by not generating information which isn't particularly related to the runtime module in the runtime module. If the deployment module 'hosts' the `quarkus-extension.yaml`, it can harvest information from the extension's build steps and build items. This is architecturally clean, unlocks a range of capabilities in the future, and is highly disruptive in the present.

== Proposal

Moving metadata generation to the 'right' place is the architecturally correct thing to do.

image::images/0002-resolution.svg[the proposed resolution]
Source: https://excalidraw.com/#json=JMERR95gRufcmnUpWy6A_,Qd9efiR40LR3wv82CxyLKg

So how do we minimise the impact?
What needs to happen for it to work?

> * The registry reads `quarkus-extension.yaml` from the deployment model, not the runtime module

Reviewer 2: ```suggestion
* The registry reads `quarkus-extension.yaml` from the deployment module, not the runtime module
```

Reviewer 2: I'm wondering whether we should also publish the yaml as its own Maven artifact with a classifier for these use-cases (in addition to bundling it in the deployment module).

Reviewer 2: Resolving binaries for use-cases limited to metadata-only processing seems like an overkill.

Author: That's a good point. I guess there's no harm in publishing it as its own thing, is there? And it maybe has a nice (tiny) side-effect of exposing some of the quarkus-build-time-intelligence into maven.

Reviewer 2: If we want to bundle the yaml in the deployment artifact then it would have to be generated as part of the deployment module build and if we want to publish it as its own artifact we have two options:
1. publish it from the deployment module (possibly with a classifier)
2. introduce a third (metadata) module that would depend on the deployment module, which would extract the yaml from the deployment module and publish it under `<quarkus-extension-artifactId>-metadata` artifactId (for example)

Reviewer 2: We should probably also confirm we want it in YAML format. The platform descriptors are still in JSON format, which is essentially the same type of metadata.

Author: I wonder if that should be a separate change/consideration? I don't think there's a strong reason to do both changes at the same time.

Author: ... but I agree using both json and yaml for similar information is slightly ugly.

Reviewer 3: fwiw, we used yaml here as intended for easier human reading. 
was even before we did platform.json to be honest. if I could go back and change things I would just have picked one format either json or yaml not both. 

about additional archive, I get the point of being nice to look it up for external tools but it means duplication and yet another location to go look.

Reviewer 2: Just to clarify, there wouldn't be an additional archive. It'd be yaml/json artifact. The advantage would be to not resolve and unpack binaries when just yaml/json is the target.

* The build of extensions' deployment module invokes a maven plugin (the existing one, or a new one)

> * (optional) Extension source code refactoring: `runtime/src/META-INF/quarkus-extension.yaml` moves to `deployment`. This eliminates one of the cross-module dependencies we'd like to avoid to stay cache-friendly.

Reviewer 2: During the "transition period", we could consider looking for `quarkus-extension.yaml` "template" in the runtime module if it is not found in the deployment one when we build the deployment module.

Author: Yes, I agree. Does https://github.com/quarkusio/quarkus/pull/40336/files/f5e181616149a2e37573296f5caeaf8bb27a76ff#diff-6fae3f8e5c70cd771717b05a93172b01c4d2a52a47f3eef96e238890d1078386R73 already cover this? I could add the 'template' word in there to make it clearer.

Reviewer 2: After the first time reading it, I thought it was about making the plugin recognize whether it's configured in the runtime or the deployment module and look for the YAML in that module whatever it is.
Perhaps, we could make more clear stating that if it is configured in the deployment module and the YAML template isn't found in the deployment module, it would look for it in the runtime module?

> The `quarkus-extension.properties` will continue to be generated in `runtime`. We should do an inventory to double-check that nothing in the `quarkus-extension.yaml` needs

Reviewer 2: ```suggestion
The `quarkus-extension.properties` will continue to be generated in the `runtime` module. We should do an inventory to double-check that nothing in the `quarkus-extension.yaml` needs
```

> * to be accessed by tools outside the registry which expect it to be in the runtime jar (IDEs?)

Reviewer 2: ```suggestion
* to be accessed by tools outside the registry which expects it to be in the runtime JAR (IDEs?)
```

* generated information which is only available to the extension plugin when it's running in the runtime module

> If so, those informations should move to `quarkus-extension.properties`.

Reviewer 2: In this case, afaiui, consumers will need an adjustment anyway and so, I'd suggest them to pull the proper artifact instead of moving metadata that is not relevant for application building to the runtime module.

Author: That's a fair point. I guess I was thinking if there was situations where the deployment module wasn't even available, because that module is generally 'invisible'. But I can't think of an example of tooling like that. I'll update the discussion and have the properties option as a contingency.

A secondary benefit of this proposal is reducing the size (slightly) of the extension's runtime artifacts.

=== Transition plan

We will need to allow a long transition period (probably one major release lifespan), in which

> * The registry looks in both the runtime and deployment jars for `META-INF/quarkus-extension.yaml`

Reviewer 2: Or, possibly, the new yaml artifact instead of the deployment JAR. To be evaluated.

(a) not every extension will have the information in the new place, so tools need to look in the old place
(b) not every tool will know about the new place, so extensions need to also package the information in the old place

== Consequences

What currently uses the `quarkus-extension.yaml` beyond the registry? Does this affect IDEs? Do users ever look into jars for the file?

== Scenarios

=== Extension owners
The ADR will affect extensions differently, depending how coupled they are to the core repo.

==== Quarkus core extensions

We will need to patch each extension's `deployment` `pom.xml` to invoke the extension plugin, and move the `src/META-INF/quarkus-extension.yaml`. We can do this any time after (or with) the plugin changes.

==== Quarkiverse extensions

We can do an auto-update using @gastaldi's script to update the deployment pom to invoke the extension plugin, but this can only be applied to extensions which are building against a Quarkus release which has these changes in.

==== Unmanaged extensions

We will need to broadcast information about the changes. We should assume that the transition will be slow for these extensions.

> === Tooling

Reviewer 2: We'll need to update platform descriptor JSON generating tooling to look for the new source of this metadata.

For platform releases, the registry will still consume these platform descriptors, so I guess there won't be much of a change for platform metadata processing on the registry side.
However, for non-platform extensions the metadata source will change and the registry will have to adjust to that, afaiu.

= Enhanced extension metadata generation

* Status: proposed
* Date: 2024-04-26

== Context and Problem Statement

> How do we know which extensions provide a dev service? At the moment, we don't, except by folk knowledge and https://quarkus.io/guides/dev-services[manually maintained lists.]

Reviewer 3: this feels to me like just one minor out of bigger issues we have.

I feel like this is only taking into consideration what is in quarkus-extension.yaml where the actual overall issue is that we never really clearly defined which phases creates/generates what files and where they are stored. 

see https://github.com/quarkusio/quarkus/issues/4539 for (probably outodate and now even more tidbits of files) that are in play.

I'm not saying all of these files are bad or good - but given we are also doing work where config metadata is being treated differently i think its worth it we look at this more holistically than just the 'need to know if extension has dev service and runtime vs deployment generation challenges'

Author: I agree there are larger discussions that we could be having, but I don't think they need to happen all at once. I'd prefer to limit the scope of this ADR to "let's move the metadata generation to a stage where more information is available".

== Considered options

- *Manual creation of metadata*. This is tedious for users and error-prone; we could not assume any manually-managed metadata was complete. It's also annoyingly non-DRY; we'd be asking extension owners to write the same information down in two places (their code, and a metadata file). It does have the advantage that it's non-breaking.
- *Running builds twice*. Because the circular dependency is a soft one, it works perfectly well to run a build to generate the deployment module's files, and then run a second build to pull those into the runtime module's metadata generation. Doing this would be a simple, non-breaking, change to release scripts. However, it is error-prone (if scripts are incorrectly run once-only, we just miss metadata), lengthens build times, breaks the Develocity by introducing hidden cross-module dependencies, and is generally Just Wrong.
- *Doing nothing*. This is cheap, and non-breaking. But it means we miss an opportunity to take advantage of the fact that, at build time, we know a lot about what extensions can do and would like to make that information available to Quarkus users.

> - *Move (most) metadata generation to the deployment module* We can eliminate the circular dependency by not generating information which isn't particularly related to the runtime module in the runtime module. If the deployment module 'hosts' the `quarkus-extension.yaml`, it can harvest information from the extension's build steps and build items. This is architecturally clean, unlocks a range of capabilities in the future, and is highly disruptive in the present.

Reviewer 3: this list is missing the consideration that both runtime and deployment has information and we then generate file a build time based on both of these.

i.e. originally the extension.yml in runtime src folder is and still is afaik treated a *input* to a generation.

could we not consider just using what is in runtime as the expected default values and then the deployment can decorate/override? Then very little need to move - its just that the lookup of the right file is done differently?

Reviewer 3: actually - i think its whats being suggested just using different words below :)

== Proposal

Moving metadata generation to the 'right' place is the architecturally correct thing to do.

image::images/0002-resolution.svg[the proposed resolution]
Source: https://excalidraw.com/#json=JMERR95gRufcmnUpWy6A_,Qd9efiR40LR3wv82CxyLKg

So how do we minimise the impact?
What needs to happen for it to work?

> * The registry reads `quarkus-extension.yaml` from the deployment module, not the runtime module. An alternate implementation would be to publish a separate `-metadata` artifact and the tooling could read that artifact.

Reviewer 3: any reason registry is called out here ?

should it not be *any* tooling (registry, our plugins, IDE etc.) that should read from the deployment module?
meaning it is no longer enough to look in the classpath but will need to lookup the matching "-deployment" ?

Author: Actually, I realised it should say 'registry upload' (https://github.com/quarkusio/jbang-catalog/blob/main/catalog_publish.java#L318). By the time it hits the registry it's pure metadata, with no enclosing jar. 

I think it’s useful to call it out the registry, since it’s the main thing which consumes and dispenses metadata. It's also the tooling we control so can directly influence. Other tooling, like IDEs, might choose to take its chances and only support the new model, or take its chances and accept the ‘old’ data.

But I'll update with the correction about the upload, and also include other tools.

* The build of extensions' deployment module invokes a maven plugin (the existing one, or a new one)

> * (optional) Extension source code refactoring:  the template `runtime/src/META-INF/quarkus-extension.yaml` moves to `deployment`. This eliminates one of the cross-module dependencies we'd like to avoid to stay cache-friendly. We want this refactoring to be optional, so if the yaml template isn't found in the deployment module, tooling would look for it in the runtime module.

Reviewer 3: this here is confusing me as i feel like what is used for generation is mixed up with the actual results.

We have two things:
input to the build (i.e. src/META-INF/quarkus-extension.yaml is read as a map and then merged with values from surrounding build) which then result in a generated output.  Today we only generate one in the runtime jar.

This proposal is saying that we should have the actual generated result in the deployment.

But that doesn't require any refactoring? just adding extra step?

Author: > This proposal is saying that we should have the actual generated result in the deployment.
> But that doesn't require any refactoring? just adding extra step?

It depends how you define refactoring. 

Having the generated result in the deployment means tooling that looks for the generated result needs to change. The generation also has to be done 'in' the deployment jar, so extension deployment modules need to invoke `quarkus-extension-maven-plugin` in their `pom.xml`. That's a pretty invasive change, and it's the one we really care about, not the final location of the generated yaml. 

(If we weren't going to change where the generation was invoked, we'd just merge https://github.com/quarkusio/quarkus/pull/40306 as-is and accept the circular dependency.)

So that leads us to the second, optional bit. If we have to change every extension's build file so that the stuff which reads file-X is elsewhere, should we also move file-X? That feels a bit cleaner, and like stuff is logically laid out. Otherwise modules contain a file whose only purpose is to be used at build-time by another file. It also means every extension ships with two versions of file-X, one in each jar. One's enriched, one's un-enriched. That's potentially confusing. 

There's also a cache implication. Source which is logically `deployment` source is living in the `runtime` module. The consequence of having source in the 'wrong' module is that anytime file-X in the runtime jar changes, the deployment jar needs to be rebuilt. Normally, we don't have to rebuild modules just because there's a new version of one of their dependencies. So it means develocity risks getting confused and caching something it shouldn't (in the very bad case), or doing unnecessary builds (*any* change to the runtime module causes us to build the deployment module).

But there's also an argument for minimising changes, and keeping file-X where it lives now in source, because it's only used for generation, and it's location doesn't actually change anything about the actual results, apart from build times.

The `quarkus-extension.properties` will continue to be generated in the runtime module. We should do an inventory to double-check that nothing in the `quarkus-extension.yaml` needs

* to be accessed by tools outside the registry which expects it to be in the runtime JAR (IDEs?)
* generated information which is only available to the extension plugin when it's running in the runtime module

> If so, the consumers will need to be updated to read the deployment module, or those informations should move to `quarkus-extension.properties` (if the consumer has no access to the deployment module).

Reviewer 3: why is it quarkus-extension.properties would be recommended here over just having it in the existing yaml ?

Author: If we keep it in the existing yaml, we have two files with the same name, and different content. That means instead of talking about something being in `quarkus-extension.yaml` we'd have to say it's in `runtime/quarkus-extension.yaml` or `deployment/quarkus-extension.yaml`. It seems like a recipe for confusion.

Reviewer 4: We will need different names if we want to do that. Or it's going to be hard to differentiate things for people gathering information about the extensions from the classpath.

Or the two files are different and clearly mention the relation between the two artifacts as part of the same extension.

Ideally, I would like two descriptors and both would contain the canonical extension they refer to e.g. `groupId:artifactId` of the runtime artifact.

Reviewer 3: I literally came to the same realization overnight @gsmet.

Basically - we already do when you have  `groupid:someextension`  ...then look for quarkus-extesnion.yaml to know its an extension  then we look up `groupid:someextension-deployment`how about we follow the same pattern here? 

name it `quarkus-extension-deployment.yaml` to know its the -deployment and it has the groupdId:artifactId:version id  in it (just like quarkus-extension.yaml) has it.

Then you merge those values or simply lookup in deployment first, runtime second.

Then the lookup mechanism is deterministic, consistent and even mostly backwards compatible (as can just keep the runtime one or have it have less values if some key/value pairs are deemed more deployment relevant)

And reason I was hinting elsewhere in PR this ADR relates to the doc config annotation processor is that I can imagine similar would make sense for the config property metainfo. that there is a combined runtime + deployment that needs gathering. 

p.s. My original comment here missed some context: quarkus-extension.properties existed when we add the .yaml file and we discussed back then if we should merge/combine it but what I recall was that the .properties file had to stay and had to stay small as it was used in a very tight loop affecting reboot/start of quarkus - thus not deemed the good place to put content. @aloubyansky might remember the details and wether still relevant.

Reviewer 2: I am not sure I see enough reasons for `quarkus-extension-deployment.yaml` atm.
I would still like to have a split between a build-related metadata used by bootstrap and the devtools related one. There will be some metadata present in both files and that's ok.
I would leave `quarkus-extension.properties` out of this. AFAIU, this ADR is about moving `quarkus-extension.yaml` to the deployment and possibly deploying it as its own artifact.

Author: @maxandersen , can we be more specific about who the "you" in "you merge those values" is? 

At the moment, we have two things that look at the metadata. 

- The ExtensionDescriptorMojo is a maven plugin that creates an enriched `quarkus-extension.yaml`, based on a source-controlled file
- The registry upload tooling takes that file and passes it to the registry. 

My proposal is that the ExtensionDescriptorMojo effectively does the merge, by running when it has access to more information. @maxandersen, is your proposal that we add a third 'thing' which does the merge?

Author: Actually, better yet, could you annotate https://github.com/quarkusio/quarkus/pull/40336/files#diff-3d9123e97ef17e972bccb4811c8a5222122bbd076824a551ed2e35bfdb8c8e74 with what your proposal is?

Reviewer 3: > I am not sure I see enough reasons for `quarkus-extension-deployment.yaml` atm.

since we are going to have extensions that have quarkus-extension.yaml in both places the main reason is to just differentiate between them without having to look in what name is in the GAV. It is not an absolute necessity - just to be very explicit about it to avoid everyone thinking the two files are equal/same because they wont be - only by chance.

" this ADR is about moving quarkus-extension.yaml to the deployment and possibly deploying it as its own artifact."  - i respectfully disagree. this is not about *moving* quarkus-extension.yaml; it is about finding way how we can capture the info that deployment + runtime parts of extension has of info. Moving it is *one* out of multiple ways that can be done - what I 'm arguing is that moving it is more problematic/harder to do than simply merging content.

And the merging can be done in multiple ways:

A) deploying a separate artifact with things merged in place
B) generating a merged file into -deployment 
C) doing the merging in the tools that need it.

...in all three cases - the platform will need ability to override values from them anyway thus I'm (for now at least) -1 on introducing a separate artifact (i.e. option A)

B - has value as it simplifies consumption but you will still require for backwards compatible reasons that tools does parts of C as you have to look for quarkus-extension.yaml in "-deployment" first , then quarkus-extension.yaml in runtime. This is where calling the merged arficact `quarkus-extension-deployment.yaml` could be useful.

C - fully backwards compatible as tools will read quarkus-extension.yaml as done today, but also read quarkus-extesnion(-deployment).yaml to lookup the key first.

A secondary benefit of this proposal is reducing the size (slightly) of the extension's runtime artifacts.

=== Transition plan

We will need to allow a long transition period (probably one major release lifespan), in which

> * The registry looks in both the runtime and deployment jars for `META-INF/quarkus-extension.yaml`, and possibly also the new metadata artifact

Reviewer 3: ```suggestion
* Any tool or code needing metadata looks in both the runtime and deployment jars for `META-INF/quarkus-extension.yaml`, and possibly also the new metadata artifact
```

Author: I think it's worth calling out the registry explicitly, since it's so central to our metadata story. We also need to make it clear that it's only "tools or code needing metadata that isn't currently getting it from the registry," which is a fairly small group.

> * The extension plugin tolerates being invoked from both `runtime` and `deployment` modules (so it's probably simplest to have one 'umbrella' plugin which handles both modules). When invoked, it will look in both the runtime and deployment modules for `src/META-INF/quarkus-extension.yaml`.

Reviewer 3: not following - shouldn't we have it be just "look in its dependency chain" so if run in runtime you listen to what the runtime has of info (as it is today afaik) and when in "-deployment" you'll need to lookup the matching runtime if any values?

Author: The key thing is the goal of ‘tolerates being invoked from either module’. Dependency chain might be how we'd implement that, in practice, but only if we'd have some extra guards, to make sure it only picks up `quarkus-extension.yaml` from the runtime module of the *same* extension, not from other extensions also in the dependency chain. 

My phrasing was careless and implied if invoked from `runtime` it might look in `deployment` which is exactly what we don't want to happen, so I'll rephrase a bit to be closer to how you've put it.

* We still package up a (less complete) `quarkus-extension.yaml` into the runtime module, so that back-level tools (such as the CLI, and maybe IDEs) can find it

image::images/0002-transition-flow.svg[a transitional process, with fallbacks]
Source::https://excalidraw.com/#json=A7UUcn3Vme98So4xp3jZ6,bemZxMtdn5vLP9MgpZtwJw

> In other words, the deployment module becomes the new 'best source', but

Reviewer 3: the way I thought about it when making extension.yaml early on was that it would be *merged* with what is in deployment.

so instead of expecting the runtime file to have everything - it is expected you read and combine/merge both.

Author: I can see the argument here. The counter-argument is that since the deployment artifact can see what’s in the runtime artifact, why not do the merging in our own mojo, rather than expect all consuming tooling to do the merging? Doing the merging in the mojo is more or less what’s being proposed by this ADR (modulo some discussion about where the source file lives).

(a) not every extension will have the information in the new place, so tools need to look in the old place

> (b) not every tool will know about the new place, so extensions need to also package the information in the old place

Reviewer 3: we should look at what those particular values are - because those values (like today) would have to be manually maintained; can't be generated as then we are back to creating build. cycles, right?

Author: I'm not sure I follow this? I'm assuming the basic source-controlled template will be the same as today, without any extra manually-maintained information.

== Consequences

> What currently uses the `quarkus-extension.yaml` beyond the registry? Does this affect IDEs? Do users ever look into jars for the file?

Reviewer 3: ```suggestion
What currently uses the `quarkus-extension.yaml` beyond the registry? Does this affect IDEs? Do users ever look into jars for the file?
```

it was always intended other tools can/should use it - but as we never really solved/clarified what this ADR is really trying to solve it is not used much.

So good thing is that we probably wont break many - bad news is - we might be surprised some code does rely on it ..in some bad way...but we should get that fixed!

== Scenarios

=== Extension owners

> The ADR will affect extensions differently, depending how coupled they are to the core repo.

Reviewer 3: can you elaborate what you mean here?

Author: We can change the pom file of every single extension in the main repo at the same time as we make changes to the `ExtensionDescriptorMojo`, so we do not need to worry about compatibility. We cannot do the same for extensions which live in other repositories.  For those extensions, we'll need to either do some automated patches, or just do a communications effort to encourage extension owners to update their pom files.

==== Quarkus core extensions

We will need to patch each extension's `deployment` `pom.xml` to invoke the extension plugin, and move the `src/META-INF/quarkus-extension.yaml`. We can do this any time after (or with) the plugin changes.

==== Quarkiverse extensions

We can do an auto-update using @gastaldi's script to update the deployment pom to invoke the extension plugin, but this can only be applied to extensions which are building against a Quarkus release which has these changes in.

==== Unmanaged extensions

We will need to broadcast information about the changes. We should assume that the transition will be slow for these extensions.

=== Tooling

==== Platform descriptor generation

The tooling which generates the platform descriptor JSON will need to be updated to look for the new source of this metadata.

==== Registry

> For platform releases, the registry consumes the generated platform descriptors, so assuming that tooling is updated, there won't be much of a change for platform metadata processing on the registry side.

Reviewer 3: afaics extensions in platform should have exact same impact as non-platform extensions.

is only when using platform to get info on these details things gets different.

we have to keep reminding ourselves that just because our main platform has a set of extensions in it there should not be anything technical stopping us or someone from outside making a platform that has a different composition.

Author: I think what this is trying to say is that the upload process to the registry is different for platform and non-platform extensions. But it's suggested text from @aloubyansky, so I'll let him comment, since he knows much more than  me about the area.
